File: leetcode/0910-nth-magical-number.cc
class Solution {
 public:
  int nthMagicalNumber(int N, int A, int B) {
    long long lcm = std::lcm((long long)A, (long long)B);
    int nInOneRound = lcm / A + lcm / B - 1;
    int times = N / nInOneRound;
    long long answerBase = times * lcm;
    long long answer = answerBase, nextA = answerBase + A,
              nextB = answerBase + B;
    int remainN = N - times * nInOneRound;
    while (remainN > 0) {
      if (nextA == nextB) {
        answer = nextA;
        nextA += A;
        nextB += B;
      } else if (nextA < nextB) {
        answer = nextA;
        nextA += A;
      } else {
        answer = nextB;
        nextB += B;
      }
      remainN--;
    }
    return answer % 1000000007;
  }
};

File: leetcode/0910-nth-magical-number.py
class Solution:
    def nthMagicalNumber(self, N: int, A: int, B: int) -> int:
        lcm=A*B/gcd(A,B)
        oneRound=lcm/A
        oneRound+=lcm/B
        oneRound-=1
        rounds=N//oneRound
        answer=int(lcm*rounds)%1000000007
        nextA=answer+A
        nextB=answer+B
        N-=oneRound*rounds
        while N>0:
            if nextA<nextB:
                answer=nextA
                nextA+=A
            else:
                answer=nextB
                nextB+=B
            N-=1
        return answer%1000000007
            
        